pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java

```diff
@@ -0,0 +1,247 @@
+                    Serializer.class);
+            this.keySerializer.configure(producerConfig.originals(), true);
+        } else {
+            this.keySerializer = keySerializer;
+            producerConfig.ignore(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG);
+        }
+
+        if (valueSerializer == null) {
+            this.valueSerializer = producerConfig.getConfiguredInstance(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+                    Serializer.class);
+            this.valueSerializer.configure(producerConfig.originals(), true);
+        } else {
+            this.valueSerializer = valueSerializer;
+            producerConfig.ignore(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
+        }
+
+        String serviceUrl = producerConfig.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG).get(0);
+        ClientConfiguration clientConf = PulsarKafkaConfig.getClientConfiguration(properties);
+        try {
+            client = PulsarClient.create(serviceUrl, clientConf);
+        } catch (PulsarClientException e) {
+            throw new RuntimeException(e);
+        }
+
+        pulsarProducerConf = new ProducerConfiguration();
+        pulsarProducerConf.setBatchingEnabled(true);
+        pulsarProducerConf.setBatchingMaxPublishDelay(1, TimeUnit.MILLISECONDS);
+
+        pulsarProducerConf.setBlockIfQueueFull(
+                Boolean.parseBoolean(properties.getProperty(ProducerConfig.BLOCK_ON_BUFFER_FULL_CONFIG, "false")));
+    }
+
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+        return send(record, null);
+    }
+
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+        org.apache.pulsar.client.api.Producer producer;
+
+        try {
+            producer = producers.computeIfAbsent(record.topic(), topic -> createNewProducer(topic));
+        } catch (Exception e) {
+            callback.onCompletion(null, e);
+            CompletableFuture<RecordMetadata> future = new CompletableFuture<>();
+            future.completeExceptionally(e);
+            return future;
+        }
+
+        Message msg = getMessage(record);
+
+        CompletableFuture<RecordMetadata> future = new CompletableFuture<>();
+        CompletableFuture<MessageId> sendFuture = producer.sendAsync(msg);
+        lastSendFuture.put(record.topic(), sendFuture);
+
+        sendFuture.thenAccept((messageId) -> {
+            future.complete(getRecordMetadata(record.topic(), msg, messageId));
+        }).exceptionally(ex -> {
+            future.completeExceptionally(ex);
+            return null;
+        });
+
+        future.handle((recordMetadata, exception) -> {
+            callback.onCompletion(recordMetadata, new Exception(exception));
+            return null;
+        });
+
+        return future;
+    }
+
+    @Override
+    public void flush() {
```

Comment:
should we introduce this api in `Pulsar-Producer` as well, because in past one of the client needed the similar functionality?